### PR TITLE
More specific push implementation.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes part of https://github.com/ethereum/solidity/issues/3688

Depends on #3693.

### CODE REVIEW

File: libsolidity/codegen/ArrayUtils.cpp [hunk 1]
```diff
@@ -774,6 +774,57 @@ void ArrayUtils::resizeDynamicArray(ArrayType const& _typeIn) const
...
+			{
+				sstore(ref, add(data, 2))
+				// Replace ref variable by new length
+				if iszero(and(data, 1)) { data := shifted_length }
+				ref := add(div(data, 2), 1)
+			}
+		})", {"ref"});
+	}
+	else
+	{
```
Comment (chriseth) [inline libsolidity/codegen/ArrayUtils.cpp]: remove

**Comment by axic** [comment]
Rebased.

**Comment by axic** [comment]
Fails with:
```
optimise_multi_stores - SolidityOptimizer
failure

ASSERTION FAILURE:
- file   : SolidityOptimizer.cpp
- line   : 606
- message: check numInstructions(m_nonOptimizedBytecode, Instruction::SSTORE) == 13 has failed [9 != 13]
```

**Comment by chriseth** [comment]
Fixed.

File: libsolidity/codegen/ArrayUtils.cpp [hunk 2]
```diff
@@ -774,6 +774,55 @@ void ArrayUtils::resizeDynamicArray(ArrayType const& _typeIn) const
...
+void ArrayUtils::incrementDynamicArraySize(ArrayType const& _type) const
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+	solAssert(_type.isDynamicallySized(), "");
+	if (!_type.isByteArray() && _type.baseType()->storageBytes() < 32)
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+
+	if (_type.isByteArray())
+	{
+		// We almost always just add 2 (length of byte arrays is shifted left by one)
```
Comment (axic) [inline libsolidity/codegen/ArrayUtils.cpp:786]: @bit-shift this is the documentation piece to be extracted as part of explaining how strings work in storage

**Review by axic — APPROVED** [review]

File: libsolidity/codegen/ArrayUtils.cpp (hunk 2 above)
Comment (erak) [reply libsolidity/codegen/ArrayUtils.cpp:786]: Thanks! I'll include this.

**Pull request merged**